cli/internal/context/context.go

```diff
@@ -57,6 +58,63 @@ func New(opts ...Option) (*Context, error) {
 	return &m, nil
 }
 
+// Splits "npm:^1.2.3" and "github:foo/bar.git" into a protocol part and a version part.
+func ParseDependencyProtocol(version string) (string, string) {
+	parts := strings.Split(version, ":")
+	if len(parts) == 1 {
+		return "", parts[0]
+	}
+
+	return parts[0], strings.Join(parts[1:], ":")
+}
+
+func IsProtocolExternal(protocol string) bool {
+	// The npm protocol for yarn by default still uses the workspace package if the workspace
+	// version is in a compatible semver range. See https://github.com/yarnpkg/berry/discussions/4015
+	// For now, we will just assume if the npm protocol is being used and the version matches
+	// its an internal dependency which matches the existing behavior before this additional
+	// logic was added.
+
+	// TODO: extend this to support the `enableTransparentWorkspaces` yarn option
+	return protocol != "" && protocol != "npm"
+}
+
+func IsReferenceToLocalPackage(packageInfos map[interface{}]*fs.PackageJSON, dependencyName, dependencyVersion string) bool {
+	pkg, localPackageFound := packageInfos[dependencyName]
+
+	if !localPackageFound {
+		return false
+	}
+
+	protocol, dependencyVersion := ParseDependencyProtocol(dependencyVersion)
+
+	if protocol == "workspace" {
+		// TODO: Since support at the moment is non-existent for workspaces that contain multiple
+		// versions of the same package name, just assume its a match and don't check the range
+		// for an exact match.
+		return true
+	} else if IsProtocolExternal(protocol) {
+		// Other protocols are assumed to be external references ("git:", "npm:", "file:" etc)
```

Comment:
This comment implies that `npm:` is an external reference, however the implementation will return `false` if `protocol == "npm"`.